horizon/facts/router.py

```diff
@@ -0,0 +1,74 @@
+from urllib.parse import urljoin
+
+from fastapi import APIRouter, Depends, Request as FastApiRequest
+from loguru import logger
+from opal_common.fetcher.providers.http_fetch_provider import HttpFetcherConfig
+from opal_common.schemas.data import DataSourceEntry
+
+from authentication import enforce_pdp_token
+from config import sidecar_config
+from facts.client import FactsClientDependency
+from startup.remote_config import get_remote_config
+
+facts_router = APIRouter(dependencies=[Depends(enforce_pdp_token)])
+
+
+def generate_opal_data_source_entry(
+    obj_type: str,
+    obj_id: str,
+    obj_key: str,
+    authorization_header: str,
+) -> DataSourceEntry:
+    remote_config = get_remote_config()
+    org_id = remote_config.context.get("org_id")
+    proj_id = remote_config.context.get("project_id")
+    env_id = remote_config.context.get("env_id")
+    url = urljoin(
+        sidecar_config.CONTROL_PLANE,
+        f"/v2/internal/opal_data/{org_id}/{proj_id}/{env_id}/{obj_type}/{obj_id}",
+    )
+
+    headers = {
+        "Authorization": authorization_header,
+    }
+    if sidecar_config.SHARD_ID:
+        headers["X-Shard-Id"] = sidecar_config.SHARD_ID
+
+    pdp_client_id = remote_config.context.get("client_id")
```

Comment:
The topic is unique for the entire lifetime of the pdp, so maybe you can store it once